RecoCTPPS/PixelLocal/src/RPixPlaneCombinatoryTracking.cc

```diff
@@ -0,0 +1,483 @@
+std::vector<std::vector<uint32_t> > RPixPlaneCombinatoryTracking::getPlaneCombinations(std::vector<uint32_t> inputPlaneList, uint32_t numberToExtract)
+{
+    uint32_t numberOfPlanes = inputPlaneList.size();
+    std::string bitmask(numberToExtract, 1); // numberToExtract leading 1's
+    bitmask.resize(numberOfPlanes, 0); // numberOfPlanes-numberToExtract trailing 0's
+    std::vector<std::vector<uint32_t> > planeCombinations;
+
+    // store the combination and permute bitmask
+    do {
+      std::vector<uint32_t> tmpPlaneList;
+      for (uint32_t i = 0; i < numberOfPlanes; ++i) { // [0..numberOfPlanes-1] integers
+        if (bitmask[i]) tmpPlaneList.push_back(inputPlaneList.at(i));
+      }
+        planeCombinations.push_back(tmpPlaneList);
+    } while (std::prev_permutation(bitmask.begin(), bitmask.end()));
+
+    return planeCombinations;
+}
+
+
+//------------------------------------------------------------------------------------------------//
+
+//This function calls it self in order to get all the possible combinations of hits having a certain selected number of planes
+//This function avoids to write for loops in cascade which will not allow to define arbitrarily the minimum number of planes to use
+//The output is stored in a map containing the vector of points and as a key the map of the point forming this vector.
+//This allows to erase the points already used for the track fit
+void RPixPlaneCombinatoryTracking::getHitCombinations(
+  const std::map<CTPPSPixelDetId, std::vector<RPixDetPatternFinder::PointInPlane> > &mapOfAllHits, 
+  std::map<CTPPSPixelDetId, std::vector<RPixDetPatternFinder::PointInPlane> >::iterator mapIterator,
+  std::map<CTPPSPixelDetId, size_t> tmpHitPlaneMap,
+  std::vector<RPixDetPatternFinder::PointInPlane> tmpHitVector,
+  std::map< std::map<CTPPSPixelDetId, size_t>, std::vector<RPixDetPatternFinder::PointInPlane> > &outputMap)
+{
+    //At this point I selected one hit per plane
+    if (mapIterator == mapOfAllHits.end())
+    {
+        outputMap[tmpHitPlaneMap] = tmpHitVector;
+        return;
+    }
+    for (size_t i=0; i<mapIterator->second.size(); i++){
+      std::map<CTPPSPixelDetId, size_t> newHitPlaneMap = tmpHitPlaneMap;
+      newHitPlaneMap[mapIterator->first] = i;
+      std::vector<RPixDetPatternFinder::PointInPlane> newVector = tmpHitVector;
+      newVector.push_back(mapIterator->second.at(i));
+      std::map<CTPPSPixelDetId, std::vector<RPixDetPatternFinder::PointInPlane> >::iterator tmpMapIterator = mapIterator;
+      getHitCombinations(mapOfAllHits, ++tmpMapIterator, newHitPlaneMap, newVector, outputMap);
+    }
+}
+
+//------------------------------------------------------------------------------------------------//
+
+std::map< std::map<CTPPSPixelDetId, size_t>, std::vector<RPixDetPatternFinder::PointInPlane> > RPixPlaneCombinatoryTracking::produceAllHitCombination(std::vector<std::vector<uint32_t> > inputPlaneCombination){
```

Comment:
here and in other places, please add line breaks to fit to at most about 100-120 chars.